lib/src/Base/Geom/LevelSetMesher.cxx

```diff
@@ -203,51 +207,74 @@ Mesh LevelSetMesher::build(const LevelSet & levelSet,
-                levelFunction.setGradient(CenteredFiniteDifferenceGradient((localVertices.getMin() - localVertices.getMax()) * epsilon + Point(dimension, epsilon), levelFunction.getEvaluation()).clone());
+                // Project the vertices not in the level set on the boundary of the level set
+                // Build the optimization problem argmin ||x - x_0||^2 such that level - f(x) >= 0, where x_0 is the current vertex
+                shiftFunction.setConstant(currentVertex);
+                ComposedFunction levelFunction(function, shiftFunction);
                 problem.setLevelFunction(levelFunction);
+                solver_.setStartingPoint(delta);
                 solver_.setProblem(problem);
-                // Try with the new gradients
+                OptimizationResult result;
+                // Here we have to catch exceptions raised by the gradient
                 try
                 {
                   solver_.run();
                   result = solver_.getResult();
                 }
                 catch(...)
                 {
-                  // There is definitely a problem with this vertex. Try a gradient-free solver
-                  Cobyla solver(solver_.getProblem());
-                  solver.setStartingPoint(solver_.getStartingPoint());
-                  LOGDEBUG(OSS() << "Problem to project point=" << currentVertex << " with solver=" << solver_ << " and finite differences for gradient, switching to solver=" << solver);
-                  solver.run();
-                  result = solver.getResult();
-                } // Even finite differences gradient failed?
-              } // Gradient failed ?
-              movedVertices.add(currentVertex + result.getOptimalPoint());
+                  LOGDEBUG(OSS() << "Problem to project point=" << currentVertex << " with solver=" << solver_ << ", using finite differences for gradient");
+                  // Here we may have to fix the gradient eg in the case of analytical functions, when Ev3 does not handle the expression.
+                  const Scalar epsilon = ResourceMap::GetAsScalar("CenteredFiniteDifferenceGradient-DefaultEpsilon");
+                  levelFunction.setGradient(CenteredFiniteDifferenceGradient((localVertices.getMin() - localVertices.getMax()) * epsilon + Point(dimension, epsilon), levelFunction.getEvaluation()).clone());
+                  problem.setLevelFunction(levelFunction);
+                  solver_.setProblem(problem);
+                  // Try with the new gradients
+                  try
+                  {
+                    solver_.run();
+                    result = solver_.getResult();
+                  }
+                  catch(...)
+                  {
+                    // There is definitely a problem with this vertex. Try a gradient-free solver
+                    Cobyla solver(solver_.getProblem());
```

Comment:
add an attribute to change the solver

Reply:
It is a fallback solver, no need to change it IMO. BTW, I don't know how to check if the user defined solver can solve an optimization problem with nonlinear constraints.